coordinator/src/main/scala/filodb.coordinator/queryplanner/SingleClusterPlanner.scala

```diff
@@ -22,8 +23,68 @@ import filodb.query.LogicalPlan._
 import filodb.query.exec.{LocalPartitionDistConcatExec, _}
 import filodb.query.exec.InternalRangeFunction.Last
 
+// scalastyle:off file.size.limit
+
 object SingleClusterPlanner {
   private val mdNoShardKeyFilterRequests = Kamon.counter("queryengine-metadata-no-shardkey-requests").withoutTags
+
+  // Is TargetSchema changing during query window.
+  private def isTargetSchemaChanging(targetSchemaChanges: Seq[TargetSchemaChange],
+                                     startMs: Long, endMs: Long): Boolean =
+    targetSchemaChanges.nonEmpty && targetSchemaChanges.exists(c => c.time >= startMs && c.time <= endMs)
+
+  // Find the TargetSchema that is applicable i.e effective for the current query window
+  private def findTargetSchema(targetSchemaChanges: Seq[TargetSchemaChange],
+                               startMs: Long, endMs: Long): Option[TargetSchemaChange] = {
+    val tsIndex = targetSchemaChanges.lastIndexWhere(t => t.time <= startMs)
+    if(tsIndex > -1)
+      Some(targetSchemaChanges(tsIndex))
+    else
+      None
+  }
+
+  /**
+   * If TargetSchema exists and all of the target-schema label filters (equals) are provided in the query,
+   * then return true.
+   * @param filters Query Column Filters
+   * @param targetSchema TargetSchema
+   * @return useTargetSchema - use target-schema to calculate query shards
+   */
+  private def useTargetSchemaForShards(filters: Seq[ColumnFilter], targetSchema: Option[TargetSchemaChange]): Boolean =
+    targetSchema match {
+      case Some(ts) if ts.schema.nonEmpty => ts.schema
+        .forall(s => filters.exists(cf => cf.column == s && cf.filter.isInstanceOf[Filter.Equals]))
+      case _ => false
+    }
+
+  /**
+   * Returns an occupied option with target schema labels iff they are defined and identical
+   *   for every leaf LogicalPlan that draws data from a shard (i.e. isn't a scalar).
+   */
+  private def getUniversalTargetSchemaLabels(lp: LogicalPlan,
+                                    tsp: TargetSchemaProvider): Option[Seq[String]] = {
+    lp match {
+      case rs: RawSeries => {
+        val rangeSelectorOpt: Option[(Long, Long)] = rs.rangeSelector match {
+          case IntervalSelector(fromMs, toMs) => Some((fromMs, toMs))
+          case _ => None
+        }
+        val targetSchemaOpt = rangeSelectorOpt.map{ case (fromMs, toMs) =>
+          val tsChanges = tsp.targetSchemaFunc(rs.filters)
+          findTargetSchema(tsChanges, fromMs, toMs).map(_.schema)
+        }
+        if (targetSchemaOpt.isDefined) targetSchemaOpt.get else None
+      }
+      case sc: ScalarPlan => Some(Seq.empty)
```

Comment:
Fixed and added comments. I see `ScalarVaryingDoublePlan` is the issue-- cases have been re-arranged so all non-leaf plans are processed before other cases.

Reply:
See my other [comment](https://github.com/filodb/FiloDB/pull/1371#discussion_r862327200) about `foo or vector(0)`-- roughly the same idea applies to `ScalarVaryingDoublePlan`. Since SVDP's should contain only one row, it doesn't make sense to push them down.

For example:
```
foo + scalar(bar{<all-target-schema-labels>, unique_timeseries_tag="tag1"})
```
where spread=1. If the scalar lies on one shard-- say, shard 0-- and `foo` lies on 0 and 1, then the scalar could only be pushed down with `foo`-0, but we'd still need to add it to `foo`-1. If the scalar lies on multiple shards, then it plainly cannot be pushed down with `foo`.